### PR TITLE
add visitor for setVariable of mysql

### LINKED ISSUE #3914: Optimization for Sharing Parser with ANTLR Visitor for MySQL
Hi folks, we would like to refactor existing parser engine with **ANTLR visitor** for better performance. It is scheduled as a long-term task and a big change for our 5.x release, **we would like to welcome anyone who is interested in it to join us for this wonderful sharding parser**! 😃 

> Take a look at the task detail below, you would learn it more. ⬇️ 

**Please follow up with your comment 📝 , if you want to join**!  What’s more, welcome your any comment and suggestion!

### PR DESCRIPTION
**Pull request by jingshanglu**
For #3914 .

Changes proposed in this pull request:
- add visitor for setVariable
- add visitor for showOther
- add test for showColumus, setVariable,showCreateTable

### CODE REVIEW

File: shardingsphere-sql-parser/shardingsphere-sql-parser-engine/src/main/java/org/apache/shardingsphere/sql/parser/sql/statement/dal/dialect/mysql/SetStatement.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.shardingsphere.sql.parser.sql.statement.dal.dialect.mysql;
+
+import org.apache.shardingsphere.sql.parser.sql.statement.dal.DALStatement;
+
+public final class SetStatement extends DALStatement {
```
Comment (terrymanu) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-engine/src/main/java/org/apache/shardingsphere/sql/parser/sql/statement/dal/dialect/mysql/SetStatement.java]: Missing java doc.

File: shardingsphere-sql-parser/shardingsphere-sql-parser-mysql/src/main/java/org/apache/shardingsphere/sql/parser/sql/segment/dal/VariableExprSegment.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+package org.apache.shardingsphere.sql.parser.sql.segment.dal;
+
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+import org.apache.shardingsphere.sql.parser.sql.segment.SQLSegment;
+
+@RequiredArgsConstructor
+@Getter
+public class VariableExprSegment implements SQLSegment {
```
Comment (terrymanu) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-mysql/src/main/java/org/apache/shardingsphere/sql/parser/sql/segment/dal/VariableExprSegment.java]: 1. Missing java doc
2. Can the class be final?

File: shardingsphere-sql-parser/shardingsphere-sql-parser-mysql/src/main/java/org/apache/shardingsphere/sql/parser/sql/segment/dal/VariableExprSegment.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+    private final int startIndex;
+
+    private final int stopIndex;
+
+    private final String variable;
+
+    private final String expr;
+
```
Comment (terrymanu) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-mysql/src/main/java/org/apache/shardingsphere/sql/parser/sql/segment/dal/VariableExprSegment.java]: Unnecessary blank lines

File: shardingsphere-sql-parser/shardingsphere-sql-parser-mysql/src/main/java/org/apache/shardingsphere/sql/parser/sql/segment/dal/VariableExprSegment.java (hunk 2 above)
Comment (terrymanu) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-mysql/src/main/java/org/apache/shardingsphere/sql/parser/sql/segment/dal/VariableExprSegment.java]: We prefer do not use abbreviation.
Please rename `expr` to `expression`

File: shardingsphere-sql-parser/shardingsphere-sql-parser-mysql/src/main/java/org/apache/shardingsphere/sql/parser/sql/segment/dal/VariableExprSegment.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+@Getter
+public class VariableExprSegment implements SQLSegment {
+
+    private final int startIndex;
+
+    private final int stopIndex;
+
+    private final String variable;
+
+    private final String expr;
```
Comment (terrymanu) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-mysql/src/main/java/org/apache/shardingsphere/sql/parser/sql/segment/dal/VariableExprSegment.java]: We prefer do not use abbreviation.
Please rename `expr` to `expression`

File: shardingsphere-sql-parser/shardingsphere-sql-parser-test/src/test/java/org/apache/shardingsphere/sql/parser/integrate/asserts/statement/dal/DALStatementAssert.java [hunk 5]
```diff
@@ -83,6 +86,8 @@ public static void assertIs(final SQLCaseAssertContext assertContext, final DALS
             ShowTableStatusStatementAssert.assertIs(assertContext, (ShowTableStatusStatement) actual, (ShowTableStatusStatementTestCase) expected);
         } else if (actual instanceof ShowIndexStatement) {
             ShowIndexStatementAssert.assertIs(assertContext, (ShowIndexStatement) actual, (ShowIndexStatementTestCase) expected);
+        } else if (actual instanceof SetStatement) {
```
Comment (terrymanu) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-test/src/test/java/org/apache/shardingsphere/sql/parser/integrate/asserts/statement/dal/DALStatementAssert.java]: `SetStatement` is a new type with `ShowXXXStatement`, it should not between `ShowIndexStatement` and `ShowStatement`, please move `SetStatement` behind `ShowStatement`

File: shardingsphere-sql-parser/shardingsphere-sql-parser-test/src/test/java/org/apache/shardingsphere/sql/parser/integrate/asserts/statement/dal/impl/SetVariableStatementAssert.java [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+package org.apache.shardingsphere.sql.parser.integrate.asserts.statement.dal.impl;
+
+import lombok.AccessLevel;
+import lombok.NoArgsConstructor;
+import org.apache.shardingsphere.sql.parser.integrate.asserts.SQLCaseAssertContext;
+import org.apache.shardingsphere.sql.parser.integrate.jaxb.domain.statement.dal.SetVariableStatementTestCase;
+import org.apache.shardingsphere.sql.parser.sql.statement.dal.dialect.mysql.SetStatement;
+
+/**
+ * set variable statement assert.
```
Comment (terrymanu) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-test/src/test/java/org/apache/shardingsphere/sql/parser/integrate/asserts/statement/dal/impl/SetVariableStatementAssert.java]: First letter of java doc should be upper case.

File: shardingsphere-sql-parser/shardingsphere-sql-parser-test/src/test/java/org/apache/shardingsphere/sql/parser/integrate/asserts/statement/dal/impl/SetVariableStatementAssert.java [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+ * @author lujingshang
+ */
+@NoArgsConstructor(access = AccessLevel.PRIVATE)
+public final class SetVariableStatementAssert {
+    
+    /**
+     * Assert describe statement is correct with expected parser result.
+     * 
+     * @param assertContext assert context
+     * @param actual actual describe statement
```
Comment (terrymanu) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-test/src/test/java/org/apache/shardingsphere/sql/parser/integrate/asserts/statement/dal/impl/SetVariableStatementAssert.java]: Incorrect java doc for `describe statement`, should be `set statement`

File: shardingsphere-sql-parser/shardingsphere-sql-parser-test/src/test/java/org/apache/shardingsphere/sql/parser/integrate/asserts/statement/dal/impl/SetVariableStatementAssert.java [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+     * @param expected expected describe statement test case
```
Comment (terrymanu) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-test/src/test/java/org/apache/shardingsphere/sql/parser/integrate/asserts/statement/dal/impl/SetVariableStatementAssert.java]: Incorrect java doc for `describe statement`, should be `set statement`

File: shardingsphere-sql-parser/shardingsphere-sql-parser-test/src/test/java/org/apache/shardingsphere/sql/parser/integrate/jaxb/domain/SQLParserTestCases.java [hunk 9]
```diff
@@ -203,6 +205,9 @@
 
     @XmlElement(name = "show-index")
     private final List<ShowIndexStatementTestCase> showIndexTestCases = new LinkedList<>();
+
+    @XmlElement(name = "set-variable")
+    private final List<SetVariableStatementTestCase> setVariableStatementTestCases = new LinkedList<>();
```
Comment (terrymanu) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-test/src/test/java/org/apache/shardingsphere/sql/parser/integrate/jaxb/domain/SQLParserTestCases.java]: The variable name `setVariableStatementTestCases` should keep consist with others, it is better to  rename as `setVariableTestCases`

**Comment by tristaZero** [comment]
@terrymanu Hi, i guess you are reviewing this PR, so do i need to give it a look?

File: shardingsphere-sql-parser/shardingsphere-sql-parser-test/src/test/java/org/apache/shardingsphere/sql/parser/integrate/jaxb/domain/SQLParserTestCases.java [hunk 10]
```diff
@@ -258,6 +263,7 @@
         putAll(showCreateTableTestCases, result);
         putAll(showTableStatusTestCases, result);
         putAll(showIndexTestCases, result);
+        putAll(setVariableStatementTestCases, result);
```
Comment (terrymanu) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-test/src/test/java/org/apache/shardingsphere/sql/parser/integrate/jaxb/domain/SQLParserTestCases.java]: Please do not interrupt `showXXXTestCases`, please move `setVariableStatementTestCases` behind `showTestCases`

File: shardingsphere-sql-parser/shardingsphere-sql-parser-test/src/test/java/org/apache/shardingsphere/sql/parser/integrate/jaxb/domain/SQLParserTestCases.java (hunk 9 above)
Comment (terrymanu) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-test/src/test/java/org/apache/shardingsphere/sql/parser/integrate/jaxb/domain/SQLParserTestCases.java]: Please do not interrupt `showXXXTestCases`, please move `setVariableStatementTestCases` behind `showTestCases`

File: shardingsphere-sql-parser/shardingsphere-sql-parser-test/src/test/java/org/apache/shardingsphere/sql/parser/integrate/jaxb/domain/segment/impl/set/ExpectedVariableExpr.java [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+
+import javax.xml.bind.annotation.XmlElement;
+
+/**
+ * Expected variable expr.
+ * 
+ * @author lujingshang
+ */
+@Getter
+public final class ExpectedVariableExpr extends AbstractExpectedSQLSegment {
```
Comment (terrymanu) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-test/src/test/java/org/apache/shardingsphere/sql/parser/integrate/jaxb/domain/segment/impl/set/ExpectedVariableExpr.java]: Please do not use abbreviation, please rename `ExpectedVariableExpr` to `ExpectedVariableExpression`

File: shardingsphere-sql-parser/shardingsphere-sql-parser-test/src/test/java/org/apache/shardingsphere/sql/parser/integrate/jaxb/domain/segment/impl/set/ExpectedVariableExpr.java [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+
+package org.apache.shardingsphere.sql.parser.integrate.jaxb.domain.segment.impl.set;
+
+import lombok.Getter;
+import org.apache.shardingsphere.sql.parser.integrate.jaxb.domain.segment.AbstractExpectedSQLSegment;
+
+import javax.xml.bind.annotation.XmlElement;
+
+/**
+ * Expected variable expr.
```
Comment (terrymanu) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-test/src/test/java/org/apache/shardingsphere/sql/parser/integrate/jaxb/domain/segment/impl/set/ExpectedVariableExpr.java]: Please do not use abbreviation, please rename `expr` to `expression`

File: shardingsphere-sql-parser/shardingsphere-sql-parser-test/src/test/java/org/apache/shardingsphere/sql/parser/integrate/jaxb/domain/segment/impl/set/ExpectedVariableExpr.java [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+ * @author lujingshang
+ */
+@Getter
+public final class ExpectedVariableExpr extends AbstractExpectedSQLSegment {
+    
+    @XmlElement(name = "variable")
+    private String variable;
+
+    @XmlElement(name = "expr")
+    private String expr;
```
Comment (terrymanu) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-test/src/test/java/org/apache/shardingsphere/sql/parser/integrate/jaxb/domain/segment/impl/set/ExpectedVariableExpr.java]: Please do not use abbreviation, please rename `expr` to `expression`

File: shardingsphere-sql-parser/shardingsphere-sql-parser-test/src/test/java/org/apache/shardingsphere/sql/parser/integrate/jaxb/domain/statement/dal/SetVariableStatementTestCase.java [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+import lombok.Setter;
+import org.apache.shardingsphere.sql.parser.integrate.jaxb.domain.segment.impl.set.ExpectedVariableExpr;
+import org.apache.shardingsphere.sql.parser.integrate.jaxb.domain.statement.SQLParserTestCase;
+
+import javax.xml.bind.annotation.XmlElement;
+import java.util.LinkedList;
+import java.util.List;
+
+/**
+ * set variable statement test case.
```
Comment (terrymanu) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-test/src/test/java/org/apache/shardingsphere/sql/parser/integrate/jaxb/domain/statement/dal/SetVariableStatementTestCase.java]: First letter of java doc should be upper case.

File: shardingsphere-sql-parser/shardingsphere-sql-parser-test/src/test/java/org/apache/shardingsphere/sql/parser/integrate/jaxb/domain/statement/dal/SetVariableStatementTestCase.java [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+ * 
+ * @author lujingshang
+ */
+@Getter
+@Setter
+public final class SetVariableStatementTestCase extends SQLParserTestCase {
+
+    @XmlElement(name = "variable-expr")
```
Comment (terrymanu) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-test/src/test/java/org/apache/shardingsphere/sql/parser/integrate/jaxb/domain/statement/dal/SetVariableStatementTestCase.java]: Please do not use abbreviation, please rename `variable-expr` to `variable-expression`

File: shardingsphere-sql-parser/shardingsphere-sql-parser-test/src/test/java/org/apache/shardingsphere/sql/parser/integrate/jaxb/domain/statement/dal/SetVariableStatementTestCase.java [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+    private final List<ExpectedVariableExpr> variableExprs = new LinkedList<>();
+
+
```
Comment (terrymanu) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-test/src/test/java/org/apache/shardingsphere/sql/parser/integrate/jaxb/domain/statement/dal/SetVariableStatementTestCase.java]: Please remove unnecessary blank lines.

File: shardingsphere-sql-parser/shardingsphere-sql-parser-test/src/test/java/org/apache/shardingsphere/sql/parser/integrate/jaxb/domain/statement/dal/ShowColumnsStatementTestCase.java [hunk 17]
```diff
@@ -17,12 +17,20 @@
...
  * Show columns statement test case.
  * 
  * @author zhangliang 
  */
 public final class ShowColumnsStatementTestCase extends SQLParserTestCase {
+
+    @XmlElement
+    private ExpectedTable table;
+
+
```
Comment (terrymanu) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-test/src/test/java/org/apache/shardingsphere/sql/parser/integrate/jaxb/domain/statement/dal/ShowColumnsStatementTestCase.java]: Please remove unnecessary blank lines.

File: shardingsphere-sql-parser/shardingsphere-sql-parser-test/src/test/resources/sql/dal/show.xml [hunk 18]
```diff
@@ -50,6 +50,15 @@
             <owner name="sharding_db" start-delimiter="`" end-delimiter="`" start-index="15" stop-index="27" />
         </table>
     </show-index>
+
```
Comment (terrymanu) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-test/src/test/resources/sql/dal/show.xml]: Please only keep single blank line.

**Comment by jingshanglu** [comment]
> The comment `Fixes #3914` will close the #3914, but the issue is not finished yet.

@terrymanu What should be written？

**Comment by terrymanu** [comment]
> > The comment `Fixes #3914` will close the #3914, but the issue is not finished yet.
> 
> @terrymanu What should be written？

Maybe `For #3914` or just `#3914`

**Comment by terrymanu** [comment]
Please pay attention for #4233

**Pull request merged**